File: neet-code-150/stacks/valid_parantheses.cpp
#include <iostream>
#include <stack>
#include <unordered_map>
using namespace std;

bool isValid(string s) {
  stack<char> result;
  unordered_map<char, char> bracketMap = {
      {')', '('},
      {']', '['},
      {'}', '{'},
  };

  for (char c : s) {
    if (bracketMap.count(c)) {

      char top_element = !result.empty() ? result.top() : '#';
      if (bracketMap[c] == top_element) {
        result.pop();
      } else {
        return false;
      }
    } else {
      result.push(c);
    }
  }

  return result.empty();
}

int main() {
  string s = "([])";
  bool result = isValid(s);
  cout << result << endl;
  return 0;
}

File: neet-code-150/stacks/valid_parantheses.py
def isValid(s: str):
    stack = []
    bracketMap = {")": "(", "]": "[", "}": "{"}

    for c in s:
        if c in bracketMap:
            if stack and stack[-1] == bracketMap[c]:
                stack.pop()
            else:
                return False

        else:
            stack.append(c)

    return len(stack) == 0


val = "()[]{}"
print(isValid(val))
